### DISCUSSION TITLE
Create one-to-many and one-to-one relation on same field.

### DISCUSSION THREAD

**Discussion by jacobchrismarsh** [prisma/prisma#20876] [Q&A] [answered]
### Question

I'm wondering how I can create a one-to-many and one-to-one relation between two tables. For example, imagine I have a `User` and `Post` table, like so:
```prisma
model User {
  id    Int @id @default(autoincrement())
  posts Post[]
}

model Post {
  id     Int  @id @default(autoincrement())
  userId Int
  user   User @relation(fields: [userId], references: [id])
}
```

Now I also want to add the ability for a user to optionally "pin" a post, and I want to create that association. How can I do it? I tried adding a `pinnedPostId` to the `User` model, but it didn't work:
```prisma
model User {
  id           Int    @id @default(autoincrement())
  posts        Post[] @relation("PostsBelongingToUser")

  pinnedPostId Int?
  pinnedPost   Post? @relation("PinnedPost", fields: [pinnedPostId], references: [id])

  @@unique([id, pinnedPostId])
}

model Post {
  id     Int   @id @default(autoincrement())
  userId Int

  user           User  @relation("PostsBelongingToUser", fields: [userId], references: [id])
  pinnedPostUser User? @relation("PinnedPost")
}
```
The error I received is:
```
Error parsing attribute "@relation": A one-to-one relation must use unique 
fields on the defining side. Either add an `@unique` attribute to the field 
`pinnedPostId`, or change the relation to one-to-many. (prisma)
```

I can't add an `@unique` attribute to `pinnedPostId` because then two users could not both have `NULL` as their `pinnedPostId`. Adding `@@unique([id, pinnedPostId])` to the `User` table did not fix it either.

What should I do here? Is there a way to do this without creating a third table that maps this relation?

**Comment by ludralph** [comment, marked as answer]
Hi @jacobchrismarsh 👋 

Thank you for raising this question.

To create a one-to-many relationship between the `User` and `Post` tables, you've already defined the schema correctly with the `User` having an array of `Post`s and the `Post` having a reference to the `User`. Now, to add the ability for a user to optionally "pin" a post, you can create an additional field in the `User` table that represents the pinned post. Since this is a one-to-one relationship (each user can have at most one pinned post), you'll modify the `User` model accordingly:

```prisma
model User {
  id           Int    @id @default(autoincrement())
  posts        Post[]
  pinnedPost   Post?  @relation("UserPinnedPost", fields: [pinnedPostId], references: [id])
  pinnedPostId Int?

  @@unique([id, pinnedPostId])
}

model Post {
  id           Int    @id @default(autoincrement())
  userId       Int
  user         User   @relation(fields: [userId], references: [id])
  pinnedByUser User[] @relation("UserPinnedPost")
}
```

In this schema:

1. The `pinnedPost` field in the `User` model represents the pinned post for that user. It's an optional field, denoted by `Post?`.
2. The `pinnedPostId` field is used to store the ID of the pinned post in the `User` model.
3. The `pinnedByUser` field in the `Post` model represents the users who have pinned that particular post. It's an array of `User` because multiple users can pin the same post.

To create a one-to-one relationship, the `@relation` directive uses a unique identifier field (`pinnedPostId` in this case) to establish the link between the `User` and `Post` models.

With this updated schema, you can interact with the pinned posts and their associations accordingly.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by jacobchrismarsh** [reply]
This definitely resolves my root issue, but I'm not sure why `pinnedByUser` has to be a `User[]`, and not a `User?`. Can you explain that more?

**Reply by ludralph** [reply]
Hi @jacobchrismarsh 

In your original post, multiple users can potentially pin the same post. Using an array `User[]` for the `pinnedByUser` field in the `Post` model makes sense. This is because the relationship is not strictly one-to-one between a post and a user. Instead, it's a many-to-many relationship where multiple users can pin the same post. When a post can be pinned by multiple users, it's a case of many users being associated with a single post. Using an array `User[]` ensures that you can store all the users who have pinned the post.

If i update the model to look like the below:
```prisma
model User {
  id           Int    @id @default(autoincrement())
  posts        Post[]
  pinnedPost   Post?  @relation("UserPinnedPost", fields: [pinnedPostId], references: [id])
  pinnedPostId Int?   @unique
}

model Post {
  id           Int   @id @default(autoincrement())
  userId       Int
  user         User  @relation(fields: [userId], references: [id])
  pinnedByUser User? @relation("UserPinnedPost")
}
```
In the updated `User` model, i've added the `@unique` directive to the `pinnedPostId` field. This enforces that each `pinnedPostId` in the `User` table must be unique. This is a valid approach if you want to ensure that a post can be pinned by only one user at most. However, this would effectively make the relationship between the `User` and `Post` tables a one-to-one relationship. Each user can have only one pinned post, and each post can be pinned by only one user. If this is the behavior you want to achieve, then this updated model is the correct one to use.

On the other hand, if you intend to allow multiple users to pin the same post, the model we discussed in the previous messages with `User[]` for `pinnedByUser` would be the appropriate representation. It all depends on the specific behavior you want to implement in your application.

I hope this helps.

**Reply by jacobchrismarsh** [reply]
But by setting `pinnedPostId` with the `@unique` attribute, then two users could not both have `NULL` as their pinned post. It removes the ability to make pinning optional. Only one user could ever have a `pinnedPostId` of `NULL`.

Is that expected?

**Reply by evangeline** [reply]
It's likely that you can have multiple users with `pinnedPostId` of `NULL`. I used CRDB + Prisma and it says here in the [docs on unique constraints](https://www.cockroachlabs.com/docs/stable/unique#details):

> You can insert NULL values into columns with the UNIQUE constraint because NULL is the absence of a value, so it is never equal to other NULL values and not considered a duplicate value. This means that it's possible to insert rows that appear to be duplicates if one of the values is NULL.

**Reply by jacobchrismarsh** [reply]
@evangeline, that's a good callout! While we don't use CRDB, we do use Postgres. Previously I had assumed Postgres wouldn't allow multiple null values in a column if there was a unique index on it. [It looks like I'm wrong about that](https://www.postgresqltutorial.com/postgresql-indexes/postgresql-unique-index/#:~:text=Introduction%20to%20PostgreSQL%20UNIQUE%20index&text=PostgreSQL%20treats%20NULL%20s%20as,creates%20a%20corresponding%20UNIQUE%20index.)! Apparently if you do want that behavior, [Postgres 15 allows](https://www.postgresql.org/docs/15/release-15.html#id-1.11.6.5.5.3.4) a `UNIQUE NULLS NOT DISTINCT` option.

**Reply by cryptiklemur** [reply]
So, i notice here in this example, prisma wont let you create/upsert a user with a pinned post because it says the pinned post is missing a userId
